README.md
# Mojo-Log-Clearable
Mojo::Log with clearable log handle


Changes
{{$NEXT}}

1.001     2021-06-16 00:03:24 EDT
  - Update IRC metadata to libera.chat

1.000     2017-11-06 21:27:23 EST
  - Add Mojo::Log::Role::Clearable, and move the implementation to there.
  - Mojo::Log::Clearable is now a simple Mojo::Log subclass that has the new
    role applied.
  - Require Mojolicious 7.42 for with_roles support.

0.003     2017-04-26 22:23:40 EDT
  - Switch installer to Module::Build::Tiny
  - Fix tests to avoid removed Mojo::Util::slurp function

0.002     2015-02-14 22:03:53-05:00 America/New_York
  - Clarify docs

0.001     2015-02-08 20:15:44-05:00 America/New_York
  - First release


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.036.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Mojo::Log with clearable log handle",
  "AUTHOR" => "Dan Book <dbook\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Mojo-Log-Clearable",
  "EXE_FILES" => [],
  "LICENSE" => "artistic_2",
  "MIN_PERL_VERSION" => "5.010001",
  "NAME" => "Mojo::Log::Clearable",
  "PREREQ_PM" => {
    "Class::Method::Modifiers" => "2.11",
    "Mojolicious" => "5.77"
  },
  "TEST_REQUIRES" => {
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "0.003",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Class::Method::Modifiers" => "2.11",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec::Functions" => 0,
  "File::Temp" => 0,
  "Mojolicious" => "5.77",
  "Test::More" => "0.88"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Mojo/Log/Role/Clearable.pm
package Mojo::Log::Role::Clearable;

use Role::Tiny;

our $VERSION = '1.002';

requires 'handle', 'path';
sub clear_handle { delete shift->{handle} }
before 'path' => sub { $_[0]->clear_handle if @_ > 1 };

1;

=head1 NAME

Mojo::Log::Role::Clearable - Role for Mojo::Log with clearable log handle

=head1 SYNOPSIS

 use Mojo::Log;
 my $log = Mojo::Log->with_roles('+Clearable')->new(path => $path1);
 $log->info($message); # Logged to $path1
 $log->path($path2);
 $log->debug($message); # Logged to $path2
 $log->path(undef);
 $log->warn($message); # Logged to STDERR
 
 # Reopen filehandle after logrotate (if logrotate sends SIGUSR1)
 $SIG{USR1} = sub { $log->clear_handle };
 
 # Apply to an existing Mojo::Log object
 $app->log->with_roles('+Clearable');

=head1 DESCRIPTION

L<Mojo::Log> is a simple logger class. It holds a filehandle once it writes to
a log, and changing L<Mojo::Log/"path"> does not open a new filehandle for
logging. L<Mojo::Log::Role::Clearable> is a role that provides a
L</"clear_handle"> method and automatically calls it when L<Mojo::Log/"path">
is modified, so the logging handle is reopened at the new path. The
L</"clear_handle"> method can also be used to reopen the logging handle after
logrotate.

=head1 ATTRIBUTES

L<Mojo::Log::Role::Clearable> augments the following attributes.

=head2 path

 $log = $log->path('/var/log/mojo.log'); # "handle" is now cleared

Log file path as in L<Mojo::Log/"path">. Augmented to call
L</"clear_handle"> when modified.

=head1 METHODS

L<Mojo::Log::Role::Clearable> composes the following methods.

=head2 clear_handle

 $log->clear_handle;

Clears L<Mojo::Log/"handle"> attribute, it will be reopened from the
L<Mojo::Log/"path"> attribute when next accessed.

=head1 AUTHOR

Dan Book, C<dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright 2015, Dan Book.

This library is free software; you may redistribute it and/or modify it undef
the terms of the Artistic License version 2.0.

=head1 SEE ALSO

L<Mojo::Log>, L<Mojo::Log::Clearable>


Build.PL
# This Build.PL for Mojo-Log-Clearable was generated by
# Dist::Zilla::Plugin::ModuleBuildTiny::Fallback 0.025
use strict;
use warnings;

my %configure_requires = (
    'Module::Build::Tiny' => '0.034',
);

my %errors = map {
    eval "require $_; $_->VERSION($configure_requires{$_}); 1";
    $_ => $@,
} keys %configure_requires;

if (!grep { $_ } values %errors)
{
    # This section for Mojo-Log-Clearable was generated by Dist::Zilla::Plugin::ModuleBuildTiny 0.015.
    use strict;
    use warnings;

    use 5.010001;
    # use Module::Build::Tiny 0.034;
    Module::Build::Tiny::Build_PL();
}
else
{
    if (not $ENV{PERL_MB_FALLBACK_SILENCE_WARNING})
    {
        warn <<'EOW'
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are running this file manually, in which case you need
to learn to first fulfill all configure requires prerequisites listed in
META.yml or META.json -- or use a cpan client to install this distribution.

You can also silence this warning for future installations by setting the
PERL_MB_FALLBACK_SILENCE_WARNING environment variable, but please don't do
that until you fix your toolchain as described above.


Errors from configure prereqs:
EOW
        . do {
            require Data::Dumper; Data::Dumper->new([ \%errors ])->Indent(2)->Terse(1)->Sortkeys(1)->Dump;
        };

        sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
    }


    # This section was automatically generated by Dist::Zilla::Plugin::ModuleBuild v6.012.
    use strict;
    use warnings;

    require Module::Build; Module::Build->VERSION(0.28);


    my %module_build_args = (
      "configure_requires" => {
        "Module::Build::Tiny" => "0.034"
      },
      "dist_abstract" => "Mojo::Log with clearable log handle",
      "dist_author" => [
        "Dan Book <dbook\@cpan.org>"
      ],
      "dist_name" => "Mojo-Log-Clearable",
      "dist_version" => "1.001",
      "license" => "artistic_2",
      "module_name" => "Mojo::Log::Clearable",
      "recursive_test_files" => 1,
      "requires" => {
        "Class::Method::Modifiers" => "2.11",
        "Mojolicious" => "7.42",
        "Role::Tiny" => "2.000002",
        "perl" => "5.010001"
      },
      "test_requires" => {
        "File::Spec" => 0,
        "File::Spec::Functions" => 0,
        "File::Temp" => 0,
        "Module::Metadata" => 0,
        "Test::More" => "0.88"
      }
    );


    my %fallback_build_requires = (
      "File::Spec" => 0,
      "File::Spec::Functions" => 0,
      "File::Temp" => 0,
      "Module::Metadata" => 0,
      "Test::More" => "0.88"
    );


    unless ( eval { Module::Build->VERSION(0.4004) } ) {
      delete $module_build_args{test_requires};
      $module_build_args{build_requires} = \%fallback_build_requires;
    }

    my $build = Module::Build->new(%module_build_args);


    $build->create_build_script;
}
